### PR TITLE
8267118: OutOfMemoryError cannot be caught as a Throwable

### PR DESCRIPTION
**Pull request by dholmes-ora**
This interesting little quirk was discovered by @iklam. During verification, the catch-type in a catch clause is normally resolved as part of being checked to see that it is a subclass of Throwable. At runtime when an exception is thrown, we also have to check if the exception being thrown is assignable to the catch-type and so handled by that catch block. For the case where the catch-type is Throwable itself, the verification subclass check trivially passes (due to a name match) without actually resolving the CP entry for the catch-type. So at runtime when the exception is thrown we have to perform the CP resolution of the catch-type. But the resolution process itself can trigger exceptions and in particular if we have thrown OutOfMemoryError, the resolution may throw a secondary OutOfMemoryError, which prevents the first from being caught by the catch block!

The fix is to force resolution of the catch-type at verification time, when it is Throwable.

To aid in debugging the original problem I've also added some new logging statements that show how we find a catch block and if we encounter further exceptions in trying to catch the exception - see example in the JBS issue.

Testing: tiers 1-3, plus the new test

Thanks,
David

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267118](https://bugs.openjdk.java.net/browse/JDK-8267118): OutOfMemoryError cannot be caught as a Throwable

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)

### Contributors
 * Ioi Lam `<iklam@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4266/head:pull/4266` \
`$ git checkout pull/4266`

Update a local copy of the PR: \
`$ git checkout pull/4266` \
`$ git pull https://git.openjdk.java.net/jdk pull/4266/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4266`

View PR using the GUI difftool: \
`$ git pr show -t 4266`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4266.diff">https://git.openjdk.java.net/jdk/pull/4266.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/oops/method.cpp [hunk 1]
```diff
@@ -226,6 +226,12 @@ void Method::print_external_name(outputStream *os, Klass* klass, Symbol* method_
 }
 
 int Method::fast_exception_handler_bci_for(const methodHandle& mh, Klass* ex_klass, int throw_bci, TRAPS) {
+  bool do_logging = log_is_enabled(Info, exceptions);
+  if (do_logging) {
+    ResourceMark rm(THREAD);
+    log_info(exceptions)("Looking for catch handler for exception of type \"%s\" in method \"%s\"",
```
Comment (coleenp) [inline src/hotspot/share/oops/method.cpp:232]: I think these should be log_debug, since this is useful for debugging but many people outside of hotspot engineers use -Xlog:exceptions.

**Review by coleenp — CHANGES_REQUESTED** [review]
This seems reasonable to me.

File: src/hotspot/share/oops/method.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/oops/method.cpp:232]: If the exception is not immediately caught, will this print a line for every method on the call stack that has a catch block? If so, that may be too verbose.

**Review by iklam — APPROVED** [review]
LGTM

File: src/hotspot/share/oops/method.cpp [hunk 2]
```diff
@@ -226,6 +226,11 @@ void Method::print_external_name(outputStream *os, Klass* klass, Symbol* method_
 }
 
 int Method::fast_exception_handler_bci_for(const methodHandle& mh, Klass* ex_klass, int throw_bci, TRAPS) {
+  if (log_is_enabled(Debug, exceptions)) {
+    ResourceMark rm(THREAD);
+    log_debug(exceptions)("Looking for catch handler for exception of type \"%s\" in method \"%s\"",
+                          ex_klass == NULL ? "NULL" : ex_klass->external_name(), mh()->name()->as_C_string());
```
Comment (coleenp) [inline src/hotspot/share/oops/method.cpp:232]: You don't need () with mh().  just mh->

File: src/hotspot/share/oops/method.cpp [hunk 3]
```diff
@@ -238,27 +243,73 @@ int Method::fast_exception_handler_bci_for(const methodHandle& mh, Klass* ex_kla
...
+
       int handler_bci = table.handler_pc(i);
       int klass_index = table.catch_type_index(i);
       if (klass_index == 0) {
+        if (log_is_enabled(Debug, exceptions)) {
+          log_debug(exceptions)("    - found catch-all handler at BCI: %d", handler_bci);
+        } else if (log_is_enabled(Info, exceptions)) {
+          ResourceMark rm(THREAD);
+          log_info(exceptions)("Found catch-all handler for exception of type \"%s\" in method \"%s\" at BCI: %d",
+                               ex_klass == NULL ? "NULL" : ex_klass->external_name(), mh()->name()->as_C_string(), handler_bci);
```
Comment (coleenp) [inline src/hotspot/share/oops/method.cpp:261]: Same here.

File: src/hotspot/share/oops/method.cpp [hunk 4]
```diff
@@ -238,27 +243,73 @@ int Method::fast_exception_handler_bci_for(const methodHandle& mh, Klass* ex_kla
...
+        }
         assert(k != NULL, "klass not loaded");
         if (ex_klass->is_subtype_of(k)) {
+          if (log_is_enabled(Debug, exceptions)) {
+            log_debug(exceptions)("    - found matching handler at BCI: %d", handler_bci);
+          } else if (log_is_enabled(Info, exceptions)) {
+            ResourceMark rm(THREAD);
+            log_info(exceptions)("Found matching handler for exception of type \"%s\" in method \"%s\" at BCI: %d",
+                                 ex_klass == NULL ? "NULL" : ex_klass->external_name(), mh()->name()->as_C_string(), handler_bci);
+          }
```
Comment (coleenp) [inline src/hotspot/share/oops/method.cpp:300]: There's so much extra code added with this logging.  It's hard to follow as it is. Could you just have the log_info level output in both this and the case above where you find the handler, here and above?  debug level logging will include info level debugging.

File: src/hotspot/share/oops/method.cpp [hunk 5]
```diff
@@ -238,27 +243,73 @@ int Method::fast_exception_handler_bci_for(const methodHandle& mh, Klass* ex_kla
...
+          log_debug(exceptions)("    - resolving catch type \"%s\"",
+                               pool->klass_name_at(klass_index)->as_C_string());
+        }
         // we know the exception class => get the constraint class
         // this may require loading of the constraint class; if verification
         // fails or some other exception occurs, return handler_bci
-        Klass* k = pool->klass_at(klass_index, CHECK_(handler_bci));
+        Klass* k = pool->klass_at(klass_index, THREAD);
+        if (HAS_PENDING_EXCEPTION) {
+        if (log_is_enabled(Debug, exceptions)) {
```
Comment (coleenp) [inline src/hotspot/share/oops/method.cpp:285]: Technically, this would be interesting enough to be log_info level.

**Review by coleenp — CHANGES_REQUESTED** [review]
Can you reduce the amount of logging code please?

File: src/hotspot/share/oops/method.cpp (hunk 5 above)
Comment (dholmes-ora) [reply src/hotspot/share/oops/method.cpp:285]: Perhaps, but again making it fit and provide sufficient contextual information is awkward and the throw is already captured at the Info level. So I would hope that if someone saw the unexpected throw logging they would use that as a cue to turn on debug logging to get more information.

Logging is really, really subjective - painfully so.

File: src/hotspot/share/oops/method.cpp (hunk 4 above)
Comment (coleenp) [reply src/hotspot/share/oops/method.cpp:300]: Logging is supposed to be non-invasive in the source code, which is why it was defined ideally to have one line like log_info(tag)(string)
Here it seems to dominate the code in this method.  Anything to make it shorter and less invasive would be better so that you can see how the code is actually finds the exception handler.  I only made a small request for simplification.  I think the formatting of the log messages wouldn't be bad if the info line is printed rather than the debug line.

[0.208s][debug][exceptions] Looking for catch handler for exception of type "java.lang.Error" in method "main"
[0.208s][debug][exceptions] - checking exception table entry for BCI 0 to 3
[0.208s][debug][exceptions] - entry covers throw point BCI 0
[0.208s][debug][exceptions] - found catch-all handler at BCI: 14

Would be:

[0.208s][debug][exceptions] Looking for catch handler for exception of type "java.lang.Error" in method "main"
[0.208s][debug][exceptions] - checking exception table entry for BCI 0 to 3
[0.208s][debug][exceptions] - entry covers throw point BCI 0
[0.198s][info     ][exceptions] Found catch-all handler for exception of type "java.lang.Error" in method "main" at BCI: 14

I don't think this looks bad at all and worth not having special if statement for debug/info logging.  That's 6 lines saved in this 96 line function.

File: src/hotspot/share/oops/method.cpp (hunk 5 above)
Comment (coleenp) [reply src/hotspot/share/oops/method.cpp:285]: ok.

File: test/hotspot/jtreg/runtime/Throwable/TestCatchThrowableOOM.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8267118
+ * @summary Test catching Throwable doesn't trigger OOME
+ * @library /test/lib
+ * @run driver TestCatchThrowableOOM
```
Comment (coleenp) [inline test/hotspot/jtreg/runtime/Throwable/TestCatchThrowableOOM.java:29]: Do you  want to run this with -Xlog:exceptions=debug to verify that the debug statements print.  I don't think you should verify them if you don't want to but make sure that the code is exercised going forward.

**Review by coleenp — CHANGES_REQUESTED** [review]

File: src/hotspot/share/oops/method.cpp (hunk 4 above)
Comment (dholmes-ora) [reply src/hotspot/share/oops/method.cpp:300]: Okay I will redo this.

File: test/hotspot/jtreg/runtime/Throwable/TestCatchThrowableOOM.java (hunk 6 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Throwable/TestCatchThrowableOOM.java:29]: The test itself doesn't need logging, the exec'd VM has all the logging enabled.

File: src/hotspot/share/oops/method.cpp [hunk 7]
```diff
@@ -238,27 +243,67 @@ int Method::fast_exception_handler_bci_for(const methodHandle& mh, Klass* ex_kla
...
       int klass_index = table.catch_type_index(i);
       if (klass_index == 0) {
+        if (log_is_enabled(Info, exceptions)) {
+          ResourceMark rm(THREAD);
+          log_info(exceptions)("Found catch-all handler for exception of type \"%s\" in method \"%s\" at BCI: %d",
+                               ex_klass == NULL ? "NULL" : ex_klass->external_name(), mh->name()->as_C_string(), handler_bci);
+        }
         return handler_bci;
       } else if (ex_klass == NULL) {
+        // Is this even possible?
```
Comment (coleenp) [inline src/hotspot/share/oops/method.cpp:263]: Good question. this seems odd - there should always be a klass.  Maybe investigate as a follow up.

File: test/hotspot/jtreg/runtime/Throwable/TestCatchThrowableOOM.java (hunk 6 above)
Comment (coleenp) [reply test/hotspot/jtreg/runtime/Throwable/TestCatchThrowableOOM.java:29]: ok.

**Review by coleenp — APPROVED** [review]
Thanks for making the changes.  It looks good!  Thank you for fixing this.

**Review by iklam — COMMENTED** [review]
Latest version LGTM.
